util/ranger/points.go

```diff
@@ -421,15 +422,48 @@ func (r *builder) buildFromNot(expr *expression.ScalarFunction) []point {
 	case ast.IsFalsity:
 		return r.buildFromIsFalse(expr, 1)
 	case ast.In:
-		rangePoints, hasNull := r.buildFromIn(expr)
+		var unsignedInt bool
+		if x, ok := expr.GetArgs()[0].(*expression.Column); ok {
+			unsignedInt = mysql.HasUnsignedFlag(x.RetType.Flag) && mysql.IsIntegerType(x.RetType.Tp)
+		}
+		inPoints, hasNull := r.buildFromIn(expr)
 		if hasNull {
 			return nil
 		}
-		retRangePoints := make([]point, 0, len(rangePoints)+2)
+		var rangePoints []point
+		// negative ranges can be directly ignored for unsigned int columns.
+		// Point `0` should be considered specially here. If the `IN` predicate contains `0`, `0` should be
+		// excluded from the result range, otherwise `0` should be included.
+		if unsignedInt {
+			rangePoints = make([]point, 0, 2+len(inPoints))
+			for i := 0; i < len(inPoints); i += 2 {
+				value := inPoints[i].value.GetInt64()
+				if value < 0 {
+					continue
+				} else if value == 0 {
```

Comment:
In this branch, the appended value should be the same as `rangePoints[i]` an  d `rangePoints[i+1]`?

Reply:
I think twice, and found that there's no need to consider `0` as a special case. I'll fix here